packages/airnode-protocol/contracts/rrp/authorizers/admin/AirnodeTokenLock.sol

```diff
@@ -0,0 +1,495 @@
+            IERC20(api3Token).transferFrom(
+                msg.sender,
+                address(this),
+                lockAmount
+            )
+        );
+
+        emit Locked(
+            _chainId,
+            _airnode,
+            _endpointId,
+            _requesterAddress,
+            msg.sender,
+            lockAmount,
+            tokenLock.whitelistCount
+        );
+    }
+
+    /// @notice Unlocks the API3 tokens for a given chain-airnode-requester-endpoint pair.
+    /// chainId-airnode-endpoint-requester pair will be unauthorized if no token lock for the pair is found.
+    /// @dev Checks whether the user (msg.sender) has already locked anything,
+    /// or if the requester address is blocked. Transfers to msg.sender the locked amount.
+    /// @param _chainId The id of the chain
+    /// @param _airnode The airnode address
+    /// @param _endpointId The endpointId
+    /// @param _requesterAddress The address of the requester for which tokens are being unlocked
+    function unlock(
+        uint256 _chainId,
+        address _airnode,
+        bytes32 _endpointId,
+        address _requesterAddress
+    ) external override isNotBlocked(_airnode, _requesterAddress) {
+        require(_chainId != 0, ERROR_ZERO_CHAINID);
+        require(_airnode != address(0), ERROR_ZERO_ADDRESS);
+        require(_requesterAddress != address(0), ERROR_ZERO_ADDRESS);
+        TokenLocks storage tokenLock = tokenLocks[_chainId][_airnode][
+            _endpointId
+        ][_requesterAddress];
+        require(
+            tokenLock.lockerToLockAmount[msg.sender] != 0,
+            ERROR_NOT_LOCKED
+        );
+
+        uint256 amount = tokenLock.lockerToLockAmount[msg.sender];
+
+        tokenLock.lockerToLockAmount[msg.sender] = 0;
+        tokenLock.whitelistCount--;
+
+        if (tokenLock.whitelistCount == 0) {
+            IRequesterAuthorizerWithManager(
+                chainIdToRequesterAuthorizerWithManager[_chainId]
+            ).setIndefiniteWhitelistStatus(
+                    _airnode,
+                    _endpointId,
+                    _requesterAddress,
+                    false
+                );
+        }
+
+        assert(IERC20(api3Token).transfer(msg.sender, amount));
+
+        emit Unlocked(
+            _chainId,
+            _airnode,
+            _endpointId,
+            _requesterAddress,
+            msg.sender,
+            amount,
+            tokenLock.whitelistCount
+        );
+    }
+
+    /// @notice Called by a block requester role to set the Block status of a airnode-requester pair.
+    /// A blocked requester can have its Locks be withdrawn into the address specified
+    /// by the blockWithdrawDestination setter. This is an irreversible action.
+    /// @dev A requester blocked on address(0) is blocked globally
+    /// @param _airnode The airnode address
+    /// @param _requesterAddress requester address
+    function blockRequester(address _airnode, address _requesterAddress)
```

Comment:
Isn't it weird that blocking a requester doesn't take the whitelisting out? It only prevents the unstake but the requester would still be able to make requests until withdrawBlocked is called.

Reply:
unfourtunately to take the whitelisting out we need the endpointId and blocking on specific endpoints seem redundant. Which is why we had to do it through withdrawBlocked.